### PR TITLE
Add Configuration State API

### LINKED ISSUE #1120: Configuration State API
Right now there's no api for when backend or client connections are in config state

Either connect the play plugin message channel api to this, or make new copy of this api for config state

### PR DESCRIPTION
**Pull request by 4drian3d**
Added events for when a player enters and finishes the configuration phase.
This, along with pull request #1224 would implement a functional way to check the State in which the player is in

blocked by #1224
resolves #1120

### CODE REVIEW

**Comment by SasaBrix** [comment]
Are there any new updates?

**Comment by 4drian3d** [comment]
> Are there any new updates?

I am waiting for some comments in order to improve this pull request, if you have something to contribute or any suggestion, you are welcome to do it

File: api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishConfigurationEvent.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import com.velocitypowered.api.proxy.Player;
+import org.jetbrains.annotations.NotNull;
+
+// TODO: Protocol State API
+/**
+ * This event is executed when the player is about to finish the Configuration state.
+ * <p>Velocity will wait for this event to finish the configuration phase on the client.</p>
+ *
+ * @param player The player who is about to complete the configuration phase
+ */
```
Comment (Timongcraft) [inline api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishConfigurationEvent.java:23]: You added the ATsince annotation to the PlayerEnterConfigurationEvent so you might want to add it here too.

File: api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishedConfigurationEvent.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+import com.velocitypowered.api.proxy.Player;
+
+/**
+ * Event executed when a player of version 1.20.2 finishes the Configuration state.
+ * <p>From this moment on, the {@link Player#protocolState()} method
+ * will return {@link ProtocolState#PLAY}.</p>
+ *
+ * @param player The player who has completed the Configuration state
+ */
```
Comment (Timongcraft) [inline api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishedConfigurationEvent.java:24]: You added the ATsince annotation to the PlayerEnterConfigurationEvent so you might want to add it here too.

**Comment by robinhickmann** [comment]
Is there anything else specifically that needs to be done before this could get merged? Just wondering if there's any plan other than just waiting on comments.

File: api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishedConfigurationEvent.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * The Velocity API is licensed under the terms of the MIT License. For more details,
+ * reference the LICENSE file in the api top-level directory.
+ */
+
+package com.velocitypowered.api.event.player.configuration;
+
+import com.velocitypowered.api.proxy.Player;
+
+/**
+ * Event executed when a player of version 1.20.2 finishes the Configuration state.
```
Comment (Luccboy) [inline api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishedConfigurationEvent.java]:
```suggestion
 * Event executed when a player of version 1.20.2 or higher finishes the Configuration state.
```

File: api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishConfigurationEvent.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+package com.velocitypowered.api.event.player.configuration;
+
+import com.velocitypowered.api.event.annotation.AwaitingEvent;
+import com.velocitypowered.api.proxy.Player;
+import org.jetbrains.annotations.NotNull;
+
+// TODO: Protocol State API
+/**
+ * This event is executed when the player is about to finish the Configuration state.
```
Comment (Luccboy) [inline api/src/main/java/com/velocitypowered/api/event/player/configuration/PlayerFinishConfigurationEvent.java:16]: You might want to specify the version (1.20.2 or higher) here too?

**Review by Luccboy — COMMENTED** [review]
This is probably unimportant, but I noticed that `Configuration phase` is sometimes written with a capital c and sometimes with a lowercase c in the JavaDocs. Maybe you want to keep the spelling consistent?

**Review by U5B — APPROVED** [review]

**Comment by Gerrygames** [comment]
I think there is currently no way to know to which server the player is currently connecting to when PlayerFinishConfigurationEvent or PlayerFinishedConfigurationEvent are called. Should probably add the ServerConnection to all configuration events.

**Comment by electronicboy** [comment]
The players current server should be switched when they enter configuration for that server, IMHO; iirc, that was missing, however (and has generally been a source of a whole bunch of other issues over the past while)

**Comment by Gerrygames** [comment]
Currently the current server isnt set until after ServerConnectedEvent has been called which is done when handling the join game packet.

**Comment by Gerrygames** [comment]
Blocking the ``PlayerFinishConfigurationEvent`` for more than ~30 seconds was not possible because the backend server would experience a timeout. I fixed this with the latest commit by not disabling auto-reading and instead modifying the ``PlayPacketQueueHandler`` to also queue inbound play packets.

**Comment by ammodev** [comment]
Is there any updates on this? I would really enjoy using this api without having to manually build velocity. Iirc theres nothing to be done here, correct?

**Review by kennytv — APPROVED** [review]

**Comment by twisti-dev** [comment]
Are there any plans to merge this?

**Comment by 4drian3d** [comment]
> Are there any plans to merge this?

Yes

**Comment by Keviro** [comment]
> Are there any plans to merge this?

There are plans, there is just no one executing them xD

**Comment by Timongcraft** [comment]
> > Are there any plans to merge this?
> 
> There are plans, there is just no one executing them xD

I think they want to bump the version first.

**Comment by ammodev** [comment]
> > > Are there any plans to merge this?
> > 
> > 
> > There are plans, there is just no one executing them xD
> 
> I think they want to bump the version first.

What else is missing to bump the version then? https://github.com/PaperMC/Velocity/milestone/6

**Comment by twisti-dev** [comment]
Hi, just wanted to follow up and see if there are any updates on merging this PR. It looks like everything is ready. Is there anything else that needs to be done before this can be merged?

Thanks!

**Pull request merged**